digest/reader.go

```diff
@@ -57,79 +56,66 @@ type FdWithDigestReader interface {
 }
 
 type fdWithDigestReader struct {
-	FdWithDigest
-
-	reader *bufio.Reader
+	fd               *os.File
+	bufReader        *bufio.Reader
+	readerWithDigest ReaderWithDigest
+	single           [1]byte
 }
 
 // NewFdWithDigestReader creates a new FdWithDigestReader.
 func NewFdWithDigestReader(bufferSize int) FdWithDigestReader {
+	bufReader := bufio.NewReaderSize(nil, bufferSize)
 	return &fdWithDigestReader{
-		FdWithDigest: newFdWithDigest(),
-		reader:       bufio.NewReaderSize(nil, bufferSize),
+		bufReader:        bufReader,
+		readerWithDigest: NewReaderWithDigest(bufReader),
 	}
 }
 
 // Reset resets the underlying file descriptor and the buffered reader.
 func (r *fdWithDigestReader) Reset(fd *os.File) {
-	r.FdWithDigest.Reset(fd)
-	r.reader.Reset(fd)
+	r.fd = fd
+	r.bufReader.Reset(fd)
+	r.readerWithDigest.Reset(r.bufReader)
 }
 
-func (r *fdWithDigestReader) readBytes(b []byte) (int, error) {
-	n, err := r.reader.Read(b)
-	if err != nil {
-		return 0, err
-	}
-	// In case the buffered reader only returns what's remaining in
-	// the buffer, recursively read what's left in the underlying reader.
-	if n < len(b) {
-		b = b[n:]
-		remainder, err := r.readBytes(b)
-		return n + remainder, err
-	}
-	return n, err
+func (r *fdWithDigestReader) Read(b []byte) (int, error) {
+	return r.readerWithDigest.Read(b)
 }
 
-func (r *fdWithDigestReader) ReadBytes(b []byte) (int, error) {
-	n, err := r.readBytes(b)
-	if err != nil && err != io.EOF {
-		return n, err
-	}
-	// If we encountered an EOF error and didn't read any bytes
-	// given a non-empty slice, we return an EOF error.
-	if err == io.EOF && n == 0 && len(b) > 0 {
-		return 0, err
-	}
-	if _, err := r.FdWithDigest.Digest().Write(b[:n]); err != nil {
-		return 0, err
-	}
-	return n, nil
+func (r *fdWithDigestReader) Fd() *os.File {
+	return r.fd
+}
+
+func (r *fdWithDigestReader) Digest() hash.Hash32 {
+	return r.readerWithDigest.Digest()
 }
 
 func (r *fdWithDigestReader) ReadAllAndValidate(b []byte, expectedDigest uint32) (int, error) {
-	n, err := r.readBytes(b)
+	n, err := r.Read(b)
 	if err != nil {
 		return n, err
 	}
-	_, err = r.reader.ReadByte()
+	_, err = r.Read(r.single[:])
```

Comment:
Maybe add a quick comment about why this is necessary.

Reply:
Sure thing.